toolchain/check/member_access.cpp

```diff
@@ -429,4 +428,81 @@ auto PerformCompoundMemberAccess(Context& context, Parse::NodeId node_id,
   return member_id;
 }
 
+// Validates that the index (required to be an IntLiteral) is valid within the
+// tuple size. Returns the index on success, or nullptr on failure.
+static auto ValidateTupleIndex(Context& context, Parse::NodeId node_id,
+                               SemIR::Inst operand_inst,
+                               SemIR::IntLiteral index_inst, int size)
+    -> const llvm::APInt* {
+  const auto& index_val = context.ints().Get(index_inst.int_id);
+  if (index_val.uge(size)) {
+    CARBON_DIAGNOSTIC(
+        TupleIndexOutOfBounds, Error,
+        "Tuple element index `{0}` is past the end of type `{1}`.", TypedInt,
+        SemIR::TypeId);
+    context.emitter().Emit(node_id, TupleIndexOutOfBounds,
+                           {.type = index_inst.type_id, .value = index_val},
+                           operand_inst.type_id());
+    return nullptr;
+  }
+  return &index_val;
+}
+
+auto PerformTupleIndex(Context& context, Parse::NodeId node_id,
+                       SemIR::InstId tuple_inst_id, SemIR::InstId index_inst_id)
+    -> SemIR::InstId {
+  tuple_inst_id = ConvertToValueOrRefExpr(context, tuple_inst_id);
+  auto tuple_inst = context.insts().Get(tuple_inst_id);
+  auto tuple_type_id = tuple_inst.type_id();
+
+  CARBON_KIND_SWITCH(context.types().GetAsInst(tuple_type_id)) {
+    case CARBON_KIND(SemIR::TupleType tuple_type): {
+      SemIR::TypeId element_type_id = SemIR::TypeId::Error;
+      auto index_node_id = context.insts().GetLocId(index_inst_id);
+      index_inst_id = ConvertToValueOfType(
+          context, index_node_id, index_inst_id,
+          context.GetBuiltinType(SemIR::BuiltinInstKind::IntType));
+      auto index_const_id = context.constant_values().Get(index_inst_id);
+      if (index_const_id == SemIR::ConstantId::Error) {
+        index_inst_id = SemIR::InstId::BuiltinError;
+      } else if (!index_const_id.is_template()) {
+        // TODO: Decide what to do if the index is a symbolic constant.
+        CARBON_DIAGNOSTIC(TupleIndexNotConstant, Error,
+                          "Tuple index must be a constant.");
```

Comment:
Is this error still produced? Looks like the test that used to trigger this case no longer does.

Reply:
this should be resolved by now

Reply:
this should be resolved by now